n_cjson.c

```diff
@@ -287,25 +284,19 @@ static Jbool parse_number(J * const item, parse_buffer * const input_buffer)
 loop_end:
     number_c_string[i] = '\0';
 
-    /* some platforms may not have locale support */
-#if !CJSON_NO_CLIB
-    number = strtod((const char*)number_c_string, (char**)&after_end);
-#else
     number = JAtoN((const char*)number_c_string, (char**)&after_end);
-#endif
     if (number_c_string == after_end) {
         return false; /* parse_error */
     }
-
     item->valuenumber = number;
 
-    /* use saturation in case of overflow */
-    if (number >= LONG_MAX) {
-        item->valueint = LONG_MAX;
-    } else if (number <= LONG_MIN) {
-        item->valueint = LONG_MIN;
+    // Saturate valueint in the case of overflow.
+    if (number >= JINTEGER_MAX) {
+        item->valueint = JINTEGER_MAX;
+    } else if (number <= JINTEGER_MIN) {
+        item->valueint = JINTEGER_MIN;
     } else {
-        item->valueint = (long int)number;
+        item->valueint = JAtoI((const char*)number_c_string);

```

Comment:
JAtoI needs to be modified to return type JINTEGER
`long int JAtoI(const char *string)`

Reply:
This gets complicated. Strap in!

When we're printing (instead of parsing), we do this:

```
if (vnum != (JNUMBER)vint) {
    JNtoA(vnum, nbuf, -1);
} else {
    JItoA(vint, nbuf);
}
```

`vnum == (JNUMBER)vint` for all integers within a certain range, depending on if JNUMBER is single or double precision. Past a certain range, each step for the floating point number (`vnum`) is greater than 1 and so you start to not be able to represent every integer exactly. See the "Precision limitations on integer values" section here: https://en.wikipedia.org/wiki/Single-precision_floating-point_format

So for certain large integers we simply can't store them exactly.

With that out of the way, for your actual question

> Do we have any way to know if the value is actually JINTEGER_MAX or an overflow?

No. Let's say we're in the `NOTE_LOWMEM` case so `JINTEGER` is `int32_t` and `JNUMBER` is `float` (single precision). Let's say we parse a JSON string with a field that's `JINTEGER_MAX + 1`. That's 2147483648. This is 2^31. Following this rule from that section of the Wikipedia page

> Integers between 2^(n) and 2^(n+1) round to a multiple of 2^(n-23).

we know that the `JNUMBER` stored when parsing will be 2^(31-23), which is 256. Indeed it gets rounded to 2147483650, which is the nearest multiple of 256. So `valueint` is 2147483647 and `valuenum` is 2147483650. Both are inaccurate, and storing a bit to indicate there was overflow doesn't help us, because neither type can precisely represent the original number. When it comes time to convert the `J` object back to a string again, the `valueint` version is what gets printed, because `vnum == (JNUMBER)vint`. If you keep increasing the value, you'll get to a point where 2147483650 is no longer the nearest multiple of 256, and we'll print `vnum` instead.

I've tried to address all this edge case stuff in the unit tests I added.

Reply:
It seems like it is possible to tell the difference, because you can check if `valuenumber` != `valueint`.